nav2_behavior_tree/CMakeLists.txt

```diff
@@ -213,12 +206,30 @@ add_library(nav2_goal_updated_controller_bt_node SHARED plugins/decorator/goal_u
 list(APPEND plugin_libs nav2_goal_updated_controller_bt_node)
 
 foreach(bt_plugin ${plugin_libs})
-  ament_target_dependencies(${bt_plugin} ${dependencies})
+  target_include_directories(${bt_plugin}
+    PRIVATE
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>")
+  target_link_libraries(${bt_plugin} PUBLIC
+    behaviortree_cpp::behaviortree_cpp
+    ${geometry_msgs_TARGETS}
+    ${nav_msgs_TARGETS}
+    ${nav2_msgs_TARGETS}
+    nav2_util::nav2_util_core
+    rclcpp::rclcpp
+    rclcpp_action::rclcpp_action
+    rclcpp_lifecycle::rclcpp_lifecycle
+    ${sensor_msgs_TARGETS}
+    tf2::tf2
+    tf2_ros::tf2_ros
+    ${std_msgs_TARGETS}
+    ${std_srvs_TARGETS}
+  )
   target_compile_definitions(${bt_plugin} PRIVATE BT_PLUGIN_EXPORT)
 endforeach()
 
-install(TARGETS ${library_name}
-                ${plugin_libs}
+install(TARGETS ${library_name} ${plugin_libs}
+  EXPORT ${library_name}
```

Comment:
plugins?

Reply:
Bah, sorry.

Reply:
Actually, no.  This is correct as-is (it helps to sleep on it).

If you look at the documentation for [install](https://cmake.org/cmake/help/latest/command/install.html#signatures), it takes multiple targets, but only a single EXPORT name.  All of the targets will be associated with that name.

Thus, downstream packages that wish to depend on this package can do so with a combination of the EXPORT name and the library they are interested in.  For instance, this will EXPORT a library called `nav2_behavior_tree::nav2_behavior_tree`, which is the library that contains the `behavior_tree_engine`.  But it will also export `nav2_behavior_tree::nav2_compute_path_to_pose_action_bt_node`, which is one of the behavior tree nodes listed in `plugin_libs`.  And it will export the rest as well.

So this should do the right thing for all of the downstream consumers that wish to use the CMake target names.

Reply:
OK. I would think having the plugins separated from the BT engine might be good since the folks using each probably have different tasks in mind which use only one or the other.

I'm not going to die on a hill about this, but seems like separating them would be good so that consumers of `behavior_tree_engine` aren't getting like 30+ and growing plugins. For instance, the BT Navigator or other's custom navigator packages based on the BT Navigator's plugin interface

Reply:
> I'm not going to die on a hill about this, but seems like separating them would be good so that consumers of `behavior_tree_engine` aren't getting like 30+ and growing plugins. For instance, the BT Navigator or other's custom navigator packages based on the BT Navigator's plugin interface

So that's the thing; they won't get all of those plugins by default.  The way that this works is that the EXPORT name acts as kind of a namespace, and the thing after the `::` is the individual library.  So linking against `nav2_behavior_tree::nav2_behavior_tree` will *only* link against `behavior_tree_engine`.  Linking against `nav2_behavior_tree::nav2_compute_path_to_pose_action_bt_node` will *only* link against that particular BT node.  And so on.

All of that said, if you prefer we can split this out to have a different namespace for the plugins.  It won't make a functional difference to downstream projects, but it may be clearer.  Just let me know if you prefer that, and I can make the change here.

Reply:
Ohhh, nice! Got it, that makes sense. I'll look over this again with that in mind, I didn't realize that.